[NoMerge] React to StringValues

benaadams opened the pull request:
https://github.com/aspnet/Common/pull/323 made a subtle change that causes `string[] myArray = new StringValues(new string[0])` to implicitly become null. 

/cc @Tratcher @rynowak

Tratcher commented:
@dougbu See http://aspnetci/viewLog.html?buildId=425455&tab=buildResultsDiv&buildTypeId=Lite_MvcTest

rynowak commented:
Is this an approved breaking change?

dougbu commented:
It was my understanding that the breaking change in `StringValues` was going to be reversed. Is that not the case @Tratcher?

benaadams commented:
Currently `StringValues` has multiple ways to represent the low counts:

Count == 0
* `null`
* `string[0]`

Count == 1
* `string`
* `string[1]`

Count == 2
* `string[2]`

This means the check `.Count == 0` becomes
```csharp
if ((_value != null ? 1 : (_values?.Length ?? 0)) == 0)
```
And the cast of a `.Count == 0` to an `string[]` can become both `string[0]` and `null`

After https://github.com/aspnet/Common/pull/323 there is a single representation for each count

Count == 0
* `null`

Count == 1
* `string`

Count == 2
* `string[2]`

There is a fast path for `.Count == 0` of `.IsNull` becoming
```csharp
if (_value == null && _values == null)
```
Obviously the inverse for `.Count > 0` being `!value.IsNull`

And the cast of a `.Count == 0` to an `string[]` can only return `null`

dougbu commented:
We were talking about a `StringValues` breaking change. Now this PR seems to create a `ValueProviderResult` breaking change (callers can no longer iterate through the values of the empty result). I don't see how either breaking change is acceptable.

benaadams commented:
> callers can no longer iterate through the values of the empty result

That should still be valid

benaadams commented:
> callers can no longer iterate through the values of the empty result

Might need these two changes also
https://github.com/aspnet/HttpAbstractions/pull/1010
https://github.com/aspnet/Common/pull/330

benaadams commented:
There were two `.Count == 0` before and things broke because not both of them were handled; now there is just one and everything is the same so only one of them needs to be handled.

e.g. the follow up change in `StringValues` https://github.com/aspnet/Common/pull/330 now tests `Assert.Empty((string[])stringValues)` as per MVC rather than the other path of `Assert.Null((string[])stringValues);` which it was doing before

benaadams commented:
Added test fix to react to other PRs https://github.com/aspnet/Mvc/pull/7520/commits/4483843653850fa462927b600bc93e38c8ada0fb

Added temp fix while waiting for other PRs https://github.com/aspnet/Mvc/pull/7520/commits/91ab7963aa9c6d8f275a21e00b5606dc4e4dfda2

Tratcher commented:
Heads up, the original change was reverted. We'll reintroduce it more cautiously by doing full universe builds before merging.

pranavkm commented:
Can this be closed?

dougbu commented:
Yes
